# Analyzing my expenses

The goal with this project is to analyse my expenses of the last two years (although I will update this project regularly) and to derive some conclusions that may help me in the future to make better financial decisions. I will try to find answers to the following questions:  
- What is my biggest expense?  
- What is my most frequent expense? How frequent is it?  
- What is the variation in monthly spending?  
- Are there any patterns? (such as spending more on the weekdays or weekends)  
- Was there any significant variation in any expense category over time?  

The app I used to record daily expenses is [Bluecoins](https://www.bluecoinsapp.com), which has has category (group) and sub-category for each expense. My from now on, I will refer to these as group and category (i.e. Housing and Mortage, or Transportation and Fuel).  

Aside from expense analysis, I will also analyze my income, as well as the resulting net earnings (= income - expenses).  

Regarding net earnings, I will try to find answers to questions such as:  
- What were the average net earnings monthly?  
- What was the variation?  
- Are there months where the net earnings are significantly higher/lower than others? What could impact this?

In [1]:
# importing libraries and the dataset

import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
transactions = pd.read_csv('transactions.csv')

## Analysing the dataset

Establishing the number of columns and rows, contained information, etc...

In [2]:
transactions.head(20)

Type                 Date Set Time                    Title   Amount  \
0   Transfer  2021-12-03 10:00:31    10:00               (Transfer)   -14.00   
1   Transfer  2021-12-03 10:00:31    10:00               (Transfer)    14.00   
2    Expense  2021-12-03 07:00:40    07:00                  Netflix   -13.99   
3    Expense  2021-12-02 20:00:05    20:00       Eating out - split   -10.39   
4    Expense  2021-12-01 17:30:57    17:30                        T    -4.50   
5    Expense  2021-12-01 10:00:15    10:00  Building administration   -31.44   
6     Income  2021-12-01 08:30:32    08:30        Capital Gains PPR    -4.87   
7   Transfer  2021-11-30 12:12:33    12:12                 Transfer  -470.56   
8   Transfer  2021-11-30 12:12:33    12:12                 Transfer   470.56   
9    Expense  2021-11-26 07:00:21    07:00                 Internet   -37.00   
10    Income  2021-11-25 12:50:41    12:50       Capital Gains T212   -12.00   
11    Income  2021-11-25 12:49:46    12:49     Capital Gains Degiro   197.00   
12    Income  2021-11-25 12:47:47    12:47        Capital Gains PPR     3.11   
13  Transfer  2021-11-25 10:59:25    10:59                 Transfer -1000.00   
14  Transfer  2021-11-25 10:59:25    10:59                 Transfer  1000.00   
15    Income  2021-11-25 10:59:24    10:59                   Salary  1912.93   
16   Expense  2021-11-25 10:00:14    10:00                 Mortgage  -250.00   
17  Transfer  2021-11-25 07:00:42    07:00                  Savings   -90.00   
18  Transfer  2021-11-25 07:00:42    07:00                  Savings    90.00   
19  Transfer  2021-11-25 07:00:22    07:00          Annual expenses   -60.00   

   Currency  Exchange Rate Category Group Name                 Category  \
0       EUR            1.0          (Transfer)               (Transfer)   
1       EUR            1.0          (Transfer)               (Transfer)   
2       EUR            1.0       Entertainment                  Netflix   
3       EUR            1.0                Food               Eating out   
4       EUR            1.0       Entertainment                        T   
5       EUR            1.0             Housing  Building Administration   
6       EUR            1.0         Investments            Capital Gains   
7       EUR            1.0          (Transfer)               (Transfer)   
8       EUR            1.0          (Transfer)               (Transfer)   
9       EUR            1.0            Personal         Internet & Phone   
10      EUR            1.0         Investments            Capital Gains   
11      EUR            1.0         Investments            Capital Gains   
12      EUR            1.0         Investments            Capital Gains   
13      EUR            1.0          (Transfer)               (Transfer)   
14      EUR            1.0          (Transfer)               (Transfer)   
15      EUR            1.0            Employer                   Salary   
16      EUR            1.0             Housing                 Mortgage   
17      EUR            1.0          (Transfer)               (Transfer)   
18      EUR            1.0          (Transfer)               (Transfer)   
19      EUR            1.0          (Transfer)               (Transfer)   

        Account Notes Labels      Status  
0    ActivoBank   NaN    NaN        None  
1         Boost   NaN    NaN        None  
2         Boost   NaN    NaN        None  
3     Splitwise   NaN    NaN  Reconciled  
4      Unibanco   NaN    NaN  Reconciled  
5    ActivoBank   NaN    NaN  Reconciled  
6           PPR   NaN    NaN  Reconciled  
7    ActivoBank   NaN    NaN  Reconciled  
8      Unibanco   NaN    NaN  Reconciled  
9         Boost   NaN    NaN        None  
10  Trading 212   NaN    NaN  Reconciled  
11       Degiro   NaN    NaN  Reconciled  
12          PPR   NaN    NaN  Reconciled  
13   ActivoBank   NaN    NaN  Reconciled  
14    Bankinter   NaN    NaN  Reconciled  
15   ActivoBank   NaN    NaN  Reconciled  
16   ActivoBank   Na

In [3]:
transactions.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2286 entries, 0 to 2285
Data columns (total 13 columns):
 #   Column               Non-Null Count  Dtype  
---  ------               --------------  -----  
 0   Type                 2286 non-null   object 
 1   Date                 2286 non-null   object 
 2   Set Time             2286 non-null   object 
 3   Title                2286 non-null   object 
 4   Amount               2286 non-null   float64
 5   Currency             2286 non-null   object 
 6   Exchange Rate        2286 non-null   float64
 7   Category Group Name  2286 non-null   object 
 8   Category             2286 non-null   object 
 9   Account              2286 non-null   object 
 10  Notes                155 non-null    object 
 11  Labels               25 non-null     object 
 12  Status               2286 non-null   object 
dtypes: float64(2), object(11)
memory usage: 232.3+ KB


### Early analysis
Looking at the above, I can see there are only two columns which contain empty values: Notes and Labels.  
All other columns contain some kind of information, with two of them standing out for containing numerical information: 'Amount' and 'Exchange rate'.

In [4]:
# information contained in 'Notes' column

nonnull_notes = transactions['Notes'].notnull()
transactions[nonnull_notes]['Notes'].head(20)

32                    Desvitalização
33                      Mx keys mini
66                      Uber Airport
74                             Luvas
75                    Bolt food Mara
76                    Bolt food Mara
102    Suporte casa de banho e velas
103    Suporte casa de banho e velas
112                    Total = 71.33
116                    Total = 71.33
127                      Mãe Mariana
137                Business calendar
173                  Jantar Andreína
192                     Quantum Park
200                        Mealheiro
201                              TGB
202                             TGB 
212                        Croissant
219                      Mistake... 
230                     Lavagem mota
Name: Notes, dtype: object

### Observations on 'Notes' column

The 'Notes' column contains information on the article that I spent money on. Since I will do the analysis based on group and category, the article itself is not important information, and I will drop this column.

In [5]:
# non-empty labels column

non_empty_labels = transactions['Labels'].notnull()
print("Labels: ")
print(transactions[non_empty_labels]['Labels'].unique())

Labels: 
['Vacation']


### Observation on the labels column

Since this column contains information on the type of expense (vacation), I won't drop it since it's important for the analysis.

In [6]:
# analysing currency, exchange rate, and account columns
print("Currencies: ")
print(transactions['Currency'].unique())
print("\n")
print("Exchange Rates: ")
print(transactions['Exchange Rate'].unique())
print("\n")
print("Accounts")
print(transactions['Account'].unique())

Currencies: 
['EUR']


Exchange Rates: 
[1.]


Accounts
['ActivoBank' 'Boost' 'Splitwise' 'Unibanco' 'PPR' 'Trading 212' 'Degiro'
 'Bankinter' 'Savings' 'Irregular Expenses' 'ticket' 'Cash' 'Moey'
 'Mintos' 'Go Parity' 'Interactive Brokers' 'Housers' 'PayPal'
 'Black Plus' 'Grupeer' 'N26']


### Conclusion on the above columns

Considering none of the columns contain essential information to my analysis, I will drop all three (it's not relevant to know which account was used to perform a purchase, only the information regarding the purchase, such as category, time, and amount).  
I will also drop the 'Title' column, considering it's only useful to input information, not necessarily the group or category of the expense, as well as the 'Status' column, since this is only useful for keeping records updated.  
The final group of columns I will use to perform my analysis will be:  
- Type (expense, income, transfer)  
- Date  
- Set Time (time of the transaction)  
- Amount  
- Category group name  
- Category  
- Labels

In [7]:
# dropping columns and renaming the dataset to 'records'

columns_to_drop = ['Title', 'Currency', 'Exchange Rate', 'Account', 'Notes', 'Status']
records = transactions.drop(columns=columns_to_drop, axis=1)
records.head(20)

Type                 Date Set Time   Amount Category Group Name  \
0   Transfer  2021-12-03 10:00:31    10:00   -14.00          (Transfer)   
1   Transfer  2021-12-03 10:00:31    10:00    14.00          (Transfer)   
2    Expense  2021-12-03 07:00:40    07:00   -13.99       Entertainment   
3    Expense  2021-12-02 20:00:05    20:00   -10.39                Food   
4    Expense  2021-12-01 17:30:57    17:30    -4.50       Entertainment   
5    Expense  2021-12-01 10:00:15    10:00   -31.44             Housing   
6     Income  2021-12-01 08:30:32    08:30    -4.87         Investments   
7   Transfer  2021-11-30 12:12:33    12:12  -470.56          (Transfer)   
8   Transfer  2021-11-30 12:12:33    12:12   470.56          (Transfer)   
9    Expense  2021-11-26 07:00:21    07:00   -37.00            Personal   
10    Income  2021-11-25 12:50:41    12:50   -12.00         Investments   
11    Income  2021-11-25 12:49:46    12:49   197.00         Investments   
12    Income  2021-11-25 12:47:47    12:47     3.11         Investments   
13  Transfer  2021-11-25 10:59:25    10:59 -1000.00          (Transfer)   
14  Transfer  2021-11-25 10:59:25    10:59  1000.00          (Transfer)   
15    Income  2021-11-25 10:59:24    10:59  1912.93            Employer   
16   Expense  2021-11-25 10:00:14    10:00  -250.00             Housing   
17  Transfer  2021-11-25 07:00:42    07:00   -90.00          (Transfer)   
18  Transfer  2021-11-25 07:00:42    07:00    90.00          (Transfer)   
19  Transfer  2021-11-25 07:00:22    07:00   -60.00          (Transfer)   

                   Category Labels  
0                (Transfer)    NaN  
1                (Transfer)    NaN  
2                   Netflix    NaN  
3                Eating out    NaN  
4                         T    NaN  
5   Building Administration    NaN  
6             Capital Gains    NaN  
7                (Transfer)    NaN  
8                (Transfer)    NaN  
9          Internet & Phone    NaN  
10            Capital Gains    NaN  
11            Capital Gains    NaN  
12            Capital Gains    NaN  
13               (Transfer)    NaN  
14               (Transfer)    NaN  
15                   Salary    NaN  
16                 Mortgage    NaN  
17               (Transfer)    NaN  
18               (Transfer)    NaN  
19               (Transfer)    NaN

In [8]:
# updating column names to make all lowercase and shorter titles

records.columns = records.columns.str.lower()
records = records.rename(columns= {'set time': 'time', 'category group name':'group'})

In [9]:
# analysing group and category columns

print('Groups: ')
print(records['group'].unique())
print('\n')

print('Categories: ')
print(records['category'].unique())

Groups: 
['(Transfer)' 'Entertainment' 'Food' 'Housing' 'Investments' 'Personal'
 'Employer' 'Others' 'Health' 'Vacations' 'Transportation' '(New Account)']


Categories: 
['(Transfer)' 'Netflix' 'Eating out' 'T' 'Building Administration'
 'Capital Gains' 'Internet & Phone' 'Salary' 'Mortgage' 'Groceries'
 'Water' 'Others' 'Appointments' 'Personal Gifts' 'Drinks' 'Coffee'
 'Electricity' 'Transportation' 'Public transport' 'Fuel' 'Clothing'
 'Cashback' 'Furniture' 'Toiletries' 'Gifts' 'Interest' 'Flights' 'Barber'
 'Sports' 'Motorcycle maintenance' 'Education' 'Restaurants'
 'Accommodation' 'Motorcycle Insurance' 'Motorcycle'
 'House Maintenance/equipment' 'Motorcycle accessories/equipment'
 'House insurance' 'Cloud Storage' 'Live Shows' 'IUC' 'E-mail' 'Medicines'
 'Bicycle' 'Fines' 'Charity' '(New Account)' 'Movies' 'Work Food' 'Gym']


In [10]:
# checking the rows with '(Transfer)' in the 'group' column

transfer_rows = records[records['group'] == '(Transfer)']
transfer_rows['amount'].sum()

-17.35000000000001

### Observations on the 'Transfer' rows

Since the value of these transfers is close to zero (a negative value will be followed by a positive one), I will drop the rows with transfer information since these aren't related with income or expenses, only movements between accounts.  
The reason why it's not exactly zero is most likely due to an accounting error or the closing of some account. Considering the amount of transactions, -17.35 is a very acceptable value and won't interfere with the analysis.

In [11]:
# updating the dataset by removing the transfer rows

records = records[records['group'] != '(Transfer)']
records.head(10)

type                 date   time   amount          group  \
2   Expense  2021-12-03 07:00:40  07:00   -13.99  Entertainment   
3   Expense  2021-12-02 20:00:05  20:00   -10.39           Food   
4   Expense  2021-12-01 17:30:57  17:30    -4.50  Entertainment   
5   Expense  2021-12-01 10:00:15  10:00   -31.44        Housing   
6    Income  2021-12-01 08:30:32  08:30    -4.87    Investments   
9   Expense  2021-11-26 07:00:21  07:00   -37.00       Personal   
10   Income  2021-11-25 12:50:41  12:50   -12.00    Investments   
11   Income  2021-11-25 12:49:46  12:49   197.00    Investments   
12   Income  2021-11-25 12:47:47  12:47     3.11    Investments   
15   Income  2021-11-25 10:59:24  10:59  1912.93       Employer   

                   category labels  
2                   Netflix    NaN  
3                Eating out    NaN  
4                         T    NaN  
5   Building Administration    NaN  
6             Capital Gains    NaN  
9          Internet & Phone    NaN  
10            Capital Gains    NaN  
11            Capital Gains    NaN  
12            Capital Gains    NaN  
15                   Salary    NaN

## Analysing the biggest expenses

In [20]:
# filtering only 'Expense' rows
expenses = records[records['type'] == 'Expense']

# gathering the highest expenses with a pivot_table, both by group and by category

pv_sum_group = expenses.pivot_table(index='group', aggfunc=np.sum).sort_values('amount')
pv_sum_category = expenses.pivot_table(index='category', aggfunc=np.sum).sort_values('amount')
pv_sum_group['percentage'] = (pv_sum_group['amount'] / pv_sum_group['amount'].sum()) * 100
pv_sum_category['percentage'] = (pv_sum_category['amount'] / pv_sum_category['amount'].sum()) * 100
pv_sum_category

amount  percentage
category                                             
Mortgage                         -6250.00   18.296531
Motorcycle                       -6115.30   17.902205
Eating out                       -2783.92    8.149773
Furniture                        -2375.47    6.954058
Groceries                        -2224.32    6.511575
Appointments                     -1591.00    4.657565
Personal Gifts                   -1240.99    3.632930
Electricity                      -1119.33    3.276777
Internet & Phone                 -1064.73    3.116939
Bicycle                          -1056.97    3.094222
Fuel                              -990.19    2.898727
Building Administration           -948.24    2.775920
Accommodation                     -783.05    2.292336
Motorcycle accessories/equipment  -723.13    2.116923
Motorcycle maintenance            -642.06    1.879595
Others                            -523.73    1.533191
Water                             -477.58    1.398089
Motorcycle Insurance              -382.17    1.118782
Gifts                             -355.63    1.041087
Clothing                          -348.27    1.019541
Education                         -297.20    0.870037
Restaurants                       -288.20    0.843690
House Maintenance/equipment       -267.18    0.782155
Public transport                  -208.59    0.610636
T                                 -144.28    0.422372
Toiletries                        -108.77    0.318418
Drinks                             -99.90    0.292452
Barber                             -97.59    0.285689
Netflix                            -95.75    0.280303
Coffee                             -88.66    0.259547
House insurance                    -80.33    0.235162
Transportation                     -65.81    0.192655
Flights                            -46.58    0.136360
Gym                                -45.96    0.134545
Medicines                          -42.57    0.124621
Sports                             -42.00    0.122953
Live Shows                         -39.62    0.115985
Fines                              -25.00    0.073186
IUC                                -19.59    0.057349
E-mail                             -15.00    0.043912
Charity                            -13.88    0.040633
Work Food                          -11.10    0.032495
Movies                             -10.69    0.031294
Cloud Storage                       -9.15    0.026786